cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -455,29 +456,50 @@ impl MachInstEmit for Inst {
                     sink,
                 );
             }
+            &Inst::MulArith32 { rd, rs1, rs2 } => {
+                let rs1 = allocs.next(rs1);
+                let rs2 = allocs.next(rs2);
+                debug_assert_eq!(rs1, b0());
+                debug_assert_eq!(rs2, e0());
+                let rd = allocs.next_writable(rd);
+                // B * E => rd
+
+                // Didn't find way to avoid stack usage, because in time of ARITH 
+                // registers are:
+                // C and D must be set to 0
+                // A and B: one must be 2 ** 32, second -- result of rs2 / 2 ** 32
+                // E: rs1 here
+                // no register to store rs2 =(
+                put_string("B :MSTORE(SP)\n", sink);
+                // E / 2 ** 32 => A
+                put_string("0 => D\n", sink);
+                put_string("0 => C\n", sink);
+                put_string("4294967296n => B\n", sink);
+                put_string("$${var _divArith = E / B}\n", sink);
+                // Intentionnaly ignore rem here because if rem != 0 mean something goes wrong
+                put_string("${_divArith} => A\n", sink);
+                put_string("E:ARITH\n", sink);
```

Comment:
I’ve been wondering for a while now, but why is it necessary to use these `var` and all of the intermediate steps, when this should basically be expressible as the following snippet?

```
E / 4294967296n => A
E: ARITH
```

Reply:
Hmm, yes, `` ${E / B} => A`` allowed. But `` ${E / 4294967296n} => A`` not allowed, so we can't free ``B`` from clobbering. Removed unnecessary ``var`` [here](https://github.com/near/wasmtime/pull/79/commits/9cabc52382570fd65474f0dd3577639b5b2c80ff)

Reply:
It does work without the `n` within `{}` though.

```
  4294967296n => C
  ${C / 4294967296} => A
  1n => B
  B :ASSERT
```

Executes just fine. Still cannot avoid clobbering B regardless as it is used as an input by the `ARITH` that follows. But my question was largely about the unnecessary complexity in the generated code, and it seems like it might not be necessary.